runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -2681,6 +2681,75 @@ J9::ARM64::TreeEvaluator::VMnewEvaluator(TR::Node *node, TR::CodeGenerator *cg)
       {
       genInitArrayHeader(node, cg, clazz, resultReg, classReg, lengthReg, zeroReg, tempReg1, isBatchClearTLHEnabled, tlhHasNotBeenCleared);
 
+      /* Here we'll update dataAddr slot for both fixed and variable length arrays. Fixed length arrays are
+       * simple as we just need to check first child of the node for array size. For variable length arrays
+       * runtime size checks are needed to determine whether to use contiguous or discontiguous header layout.
+       *
+       * In both scenarios, arrays of non-zero size use contiguous header layout while zero size arrays use
+       * discontiguous header layout.
+       */
+      TR::Register *offsetReg = tempReg1;
+      TR::Register *firstDataElementReg = tempReg2;
+      TR::MemoryReference *dataAddrSlotMR = NULL;
+
+      if (isVariableLength && TR::Compiler->om.compressObjectReferences())
+         {
+         /* We need to check lengthReg (array size) at runtime to determine correct offset of dataAddr field.
+          * Here we deal only with compressed refs because dataAddr offset for discontiguous and contiguous
+          * arrays is the same in full refs.
+          */
+         if (comp->getOption(TR_TraceCG))
+            traceMsg(comp, "Node (%p): Dealing with compressed refs variable length array.\n", node);
+
+         TR_ASSERT_FATAL_WITH_NODE(node,
+            (fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()) == 8,
+            "Offset of dataAddr field in discontiguous array is expected to be 8 bytes more than contiguous array. "
+            "But was %d bytes for discontigous and %d bytes for contiguous array.\n",
+            fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField());
+
+         // Since array size is capped at 32 bits, we don't need to check all 64 bits of lengthReg.
+         // NOTE: ARM uses inverted carry flag for subtraction operations so 0 here indicates a borrow.
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::subsimmw, node, offsetReg, lengthReg, 1);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::adcx, node, offsetReg, zeroReg, zeroReg);
```

Comment:
I would use the following instructions for this operation (not tested):
```
generateCompareImmInstruction(cg, node, lengthReg, 0);
generateCSetInstruction(cg, node, offsetReg, TR::CC_NE);
```

Reply:
@knn-k What source register does `generateCSetInstruction` use? The function only accepts only one register, the target register [1]. We can't use `offsetReg` as the source register because it doesn't get cleared before being used here.

[1] https://github.com/eclipse/omr/blob/a99c1b640cb9b287aa334e4eba99945a1957227e/compiler/aarch64/codegen/GenerateInstructions.cpp#L532 "generateCSetInstruction"

Reply:
`generateCSetInstruction()` does not take any source registers.  AArch64 `cset` (conditional set) instruction sets 1 or 0 based on the flag that was set by the previous `cmp` instruction.

```
cmp lengthReg, #0
cset offsetReg, ne // offsetReg=1 if lengthReg != 0, otherwise offsetReg=0
```

Reply:
ah ok that makes sense. I have updated my changes. Going to run a personal build to test the updated sequence.